### PR TITLE
Performance: reuse search tree in MCTS and use cache for NN

### PR DESCRIPTION
**Pull request by barakugav**

### CODE REVIEW

File: train/train_process.py [hunk 1]
```diff
@@ -90,7 +118,9 @@ def run_training_loop(self):
             games_gen_idx += 1
 
             # Train latest model from training data
-            latest_model, _metrics = self._train(latest_model, games_dir)
+            lr = lr_scheduler.get_lr(
+                iter_num * self.cfg["training"]["iteration_data_entries"])
```
Comment (poja) [inline train/train_process.py]: What exactly is counted here?

File: src/game/mcts.rs [hunk 2]
```diff
@@ -23,14 +34,13 @@ struct MCTSNode<Position: GamePosition> {
     simulations_n: u32,
 
     /// This is the variable w from UCT formula
-    /// Float because could be half for games with ties
     score_w: f32,
 }
 
-impl<Position: GamePosition> MCTSNode<Position> {
-    pub fn from_position(position: Position, init_score: f32) -> Self {
+impl<Move: GameMove> MCTSEdge<Move> {
```
Comment (poja) [inline src/game/mcts.rs:40]: General question about this commit - why is it better to have the info on the edges?
1. If the search "tree" really is a tree, then it makes no real difference, right?
The only difference is that the current/old implementation is not intuitive about the meaning of a node's `score` relative to that node's "player to play"? 

2. If we allow the search tree to be a DAG:
* Remind me - is this really a possible future scenario?
* In this scenario, would we want the info to be on edges or nodes? Why? (My first intuition is nodes)

**Review by poja — CHANGES_REQUESTED** [review]

File: train/config.json [hunk 3]
```diff
@@ -21,7 +28,15 @@
...
         "iteration_data_entries": 1024,
         "batch_size": 32,
+        "learning_rate": [
+            [40960, 0.01],
+            [61440, 0.001],
+            [       0.0001]
+        ],
         "compare": {
+            "temperature_policy": [
+                [       0.0]
```
Comment (poja) [inline train/config.json:38]: Doesn't this cause all the games to be the same? We can different games to be compared

File: train/train_process.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (barakugav) [reply train/train_process.py]: Number of data entries. Not sure what alphazero does. Batching doesnt affect this.
We can discuss this

File: src/chess/net/net_tf_two_headed.rs [hunk 4]
```diff
@@ -1,17 +1,24 @@
...
 use crate::game::net::TwoHeadedNetBase;
 
 pub struct TwoHeadedNet {
-    base: TwoHeadedNetBase,
+    base: TwoHeadedNetBase<ChessGame>,
 }
 
 impl TwoHeadedNet {
     pub fn new(model_path: &String) -> Self {
+        Self::with_cache(model_path, Arc::new(ValueFuncCache::new(1)))
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/chess/net/net_tf_two_headed.rs:16]: Agree, rust abstraction is not the best... Will do it

File: src/game/self_play_cmd.rs [hunk 5]
```diff
@@ -27,65 +30,87 @@ struct SelfPlayArgs {
     explore_factor: f32,
     #[clap(long, default_value = "1")]
     threads: u32,
+    #[clap(long, default_value = "100000")]
+    cache_size: usize,
 }
 
-pub trait ValueFunctionBuilder<Game: IGame>: Sync + Send {
-    fn new_value_func(&self, model_path: &String) -> Box<dyn ValueFunction<Game>>;
+pub trait IModelBuilder<Game: IGame>: Sync + Send {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/self_play_cmd.rs:43]: If if the interface accept model name, it is already specific for NN rather than any value function.
All this abstraction is clumsy, but i didnt find easier way in rust

File: src/game/cache.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::RwLock;
+
+struct PositionCache<Game: IGame> {
+    map: HashMap<Game::Position, (f32, Vec<(Game::Move, f32)>)>,
+    deque: VecDeque<Game::Position>,
+}
+
+pub struct ValueFuncCache<Game: IGame> {
+    lock: RwLock<PositionCache<Game>>,
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/cache.rs:13]: There is the same cache for any model, shared between threads.
This is managed by the builder

File: src/bin/chess_cmd_vs_stockfish.rs [hunk 7]
```diff
@@ -13,9 +13,9 @@ fn color_to_str(c: Option<GameColor>) -> String {
 }
 
 fn main() {
-    let mut player1 = ChessPlayerCmd::new();
-    let value_func = Box::new(StockfishNet::new());
-    let mut player2 = MCTSPlayer::new_custom(100000, 1.41421, value_func);
+    let mut player1 = ChessPlayerCmd {};
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/bin/chess_cmd_vs_stockfish.rs:16]: Clippy recommendation is against "new()" with no argument, as the default constructor "MyClass {}" is still visible to the user, and it might skip some critical code in "new()".
The recommendation is to implement some trait that controls the behavior of "MyClass {}".
In such cases in our code, the "new()" didnt actually did anything, so it was easier to just remove it and use the default constructor

File: src/game/mcts.rs [hunk 8]
```diff
@@ -356,6 +383,41 @@ impl<Game: IGame> MCTSPlayer<Game> {
...
+            .edges(node_id)
+            .map(|edge| edge.target())
+            .collect_vec();
+        if children.len() < 2 {
+            return;
+        }
+
+        /* The Dirichlet implementation seems to return NaNs and INFs sometimes. */
+        /* Keep drawing random noises until valid values are achieved */
+        let dist = Dirichlet::new_with_size(self.prior_noise_alpha, children.len()).unwrap();
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/mcts.rs]: The Dirichlet distribution operate on multiple  dimensions, denoted 'size' in the rust docs.
I this the paper ment "Dir([0.03]*move_num)".
This also match Lc0 impll

File: src/ttt/net/two_headed_net.rs [hunk 9]
```diff
@@ -22,7 +22,7 @@ impl TwoHeadedNet {
     }
 }
 
-impl ValueFunction<TttGame> for TwoHeadedNet {
+impl<const CPU: bool> ValueFunction<TttGame> for TwoHeadedNet<CPU> {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/ttt/net/two_headed_net.rs:25]: I guess that can work.
I kind of like the use of generics here, as all the 'if's are known at compile time.
Maybe we can get the env variable at the rust main and propagate it from there using generics.
But in general i would like to avoid env variable, its hard to follow who set them and where are they used

File: src/game/mcts.rs (hunk 2 above)
Comment (barakugav) [reply src/game/mcts.rs:40]: The change is only for intuitive benefits, but it does complicate the code a little bit, I think it worth it.
DAG will complicate backpropagation, and will require global table of pos->node, which might be worth it
I guess in such case we will prefer data on nodes

File: train/config.json (hunk 3 above)
Comment (barakugav) [reply train/config.json:38]: Well shit...
Dir noise?
If we want to compare them in an accurate way we should avoid both temperature and Dir noise. The paper says temperature is zero but nothing about the noise.
Maybe a better approch will be to use random position (no idea how to generate) with no random at all

Comment (poja) [reply train/config.json:38]: Maybe they can both play with high noise/temperature only in the first few turns (very few, e.g. 3), then the rest of the game is 0 noise/temp

Comment (poja) [reply train/config.json:38]: Also I think we should add a consistency (no-random) test in the case of temperature=noise=0

File: src/game/mcts.rs (hunk 2 above)
Comment (poja) [reply src/game/mcts.rs:40]: Sababa

File: src/ttt/net/two_headed_net.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/ttt/net/two_headed_net.rs:25]: If we propagate in the same way then I don't see a reason to use env variable instead of argument. Yes I agree about compile time coolness, and that generally passing information using env variables is confusing, but in this case it really is a potentially global trait of the system (has or doesn't have GPU).
Kitzur if you prefer the current method then let's keep it, both seem fine to me

File: src/chess/net/net_tf_two_headed.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/chess/net/net_tf_two_headed.rs:16]: Hmm, I think I understand why Rust is tricky here...
Maybe the TwoHeadedNet can have a `Optional<ValueFuncCache>`?
Or alternatively a `Box<IValueFuncCache>` - does Box work with Traits?

File: src/bin/chess_cmd_vs_stockfish.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/bin/chess_cmd_vs_stockfish.rs:16]: Ahh okay, it takes a bit of getting used to (my first thought was - okay, then make the `MyClass {}` invisible from outside! But that's not how it works, it seems like initiating a class "directly" is legit "by definition")
BTW was is the trait you mentioned?

File: train/config.json (hunk 3 above)
Comment (poja) [reply train/config.json:38]: What noise??

image

File: src/game/common.rs [hunk 10]
```diff
@@ -55,7 +58,7 @@ pub trait GamePosition: Clone + Copy + Eq + Hash {
     fn print(&self);
 }
 
-pub trait GameMove: Clone + Copy + Eq + Eq + Hash + Display + Debug {
+pub trait GameMove: Clone + Copy + Eq + Eq + Hash + Display + Debug + Send + Sync {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/common.rs:63]: done

[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/common.rs:63]: done

File: src/chess/net/net_tf_two_headed.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (barakugav) [reply src/chess/net/net_tf_two_headed.rs:16]: Didn't thought about Optional, i think its a good alternative. done

[the remark this answers is not in the document]
Comment (barakugav) [reply src/chess/net/net_tf_two_headed.rs:16]: Didn't thought about Optional, i think its a good alternative. done

File: src/game/cache.rs [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+                let cached_val = cached_val.unwrap().clone();
+                assert!(computed_val == cached_val);
+                return cached_val;
+            }
+
+            // Remove oldest cached elements if needed
+            while cache.deque.len() >= self.max_size {
+                let pos = *cache.deque.front().unwrap();
+                cache.map.remove(&pos);
+                cache.deque.pop_front();
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/cache.rs:67]: done

File: src/game/self_play.rs [hunk 12]
```diff
@@ -254,3 +268,40 @@ impl<Game: IGame> SelfPlayWorker<Game> {
         return Ok(());
     }
 }
+
+struct TemperatureSchedular {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/self_play.rs:263]: done

File: src/game/self_play.rs [hunk 13]
```diff
@@ -254,3 +268,40 @@ impl<Game: IGame> SelfPlayWorker<Game> {
...
     }
 }
+
+struct TemperatureSchedular {
+    temperatures: Vec<(u32, f32)>,
+    last_temperature: f32,
+}
+
+impl TemperatureSchedular {
+    fn from_str(s: &String) -> Self {
```
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/self_play.rs]: done

File: src/bin/chess_cmd_vs_stockfish.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (barakugav) [reply src/bin/chess_cmd_vs_stockfish.rs:16]: the 'Default' trait

File: src/game/self_play_cmd.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/game/self_play_cmd.rs:43]: So maybe NNetworkBuilder? Whatever you prefer

File: src/game/cache.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/game/cache.rs:13]: But there are two different builders, one for each player, so that means two caches?

[the remark this answers is not in the document]
Comment (poja) [reply src/game/cache.rs:13]: Never mind

File: src/game/mcts.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (poja) [reply src/game/mcts.rs]: Okay, not trivial but yes I am convinced from their `ApplyDirichletNoise` plus Wikipedia: Dirichlet distribution

File: src/game/self_play_cmd.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (barakugav) [reply src/game/self_play_cmd.rs:43]: done
